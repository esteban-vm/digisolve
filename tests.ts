import type { ReactElement } from 'react'
import { create, type ReactTestRendererJSON, type TestRendererOptions } from 'react-test-renderer'

const customCreate = (ui: ReactElement, options?: TestRendererOptions) => {
  return create(ui, options).toJSON() as ReactTestRendererJSON
}

export { customCreate as create }
export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'
