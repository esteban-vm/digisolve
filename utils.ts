export const formatPrice = (price: number) => {
  return new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: 'USD',
    maximumFractionDigits: 0,
  }).format(price)
}
