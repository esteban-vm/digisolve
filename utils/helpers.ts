import Filter from 'bad-words'

const filter = new Filter()
/** Checks if a message does not contain profane language */
export const isClean = (message: string) => !filter.isProfane(message)

const formatter = new Intl.NumberFormat('en-US', { style: 'currency', currency: 'USD', maximumFractionDigits: 0 })
/** Formats a price in number to USD currency */
export const formatPrice = (price: number) => formatter.format(price)

/** Gets the name of an icon from kebab case to pascal case */
export const getIconName = <T extends string>(name: T) => {
  return name.replace(/(^\w|-\w)/g, (name) => name.replace(/-/, '').toUpperCase())
}
