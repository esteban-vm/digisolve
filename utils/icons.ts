import { fas } from '@fortawesome/free-solid-svg-icons'
import { fab } from '@fortawesome/free-brands-svg-icons'

// Issue: https://github.com/FortAwesome/Font-Awesome/issues/19348
// eslint-disable-next-line @typescript-eslint/no-var-requires
const { library } = require('@fortawesome/fontawesome-svg-core')
library.add(fas, fab)
