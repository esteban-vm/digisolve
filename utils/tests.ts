import type { ReactElement } from 'react'
import { create, type TestRendererOptions } from 'react-test-renderer'

const customCreate = (ui: ReactElement, options?: TestRendererOptions) => {
  return create(ui, options).toJSON()
}

export { customCreate as create }
export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'
