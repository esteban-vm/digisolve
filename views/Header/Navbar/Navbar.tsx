import type { Component } from '@/types'
import Image from 'next/image'
import Link from 'next/link'
import { logo } from '@/images'
import { styled } from '@/styles'

export const navLinks = ['Home', 'Contact', 'About']

const NavbarComponent: Component = (props) => {
  return (
    <nav {...props}>
      <Image src={logo} alt='Digisolve logo' className='navbar__image' />
      <ul className='navbar__list'>
        {navLinks.map((link, index) => (
          <li key={index} className='navbar__item'>
            <Link href='#' className='navbar__link'>
              {link}
            </Link>
          </li>
        ))}
      </ul>
    </nav>
  )
}

const Navbar = styled(NavbarComponent)`
  .navbar {
    &__image {
      height: 120px;
      width: auto;
      float: left;
    }

    &__list {
      list-style: none;
      float: right;
      margin-top: 70px;
    }

    &__item {
      display: inline-block;
      margin-right: 40px;
    }

    &__link {
      :link,
      :visited {
        text-decoration: none;
        text-transform: uppercase;
        color: var(--color-white);
        outline: none;
        transition: border-bottom 100ms;
      }

      :hover,
      :active {
        border-bottom: 2px solid var(--color-primary);
      }
    }
  }
`

export default Navbar
