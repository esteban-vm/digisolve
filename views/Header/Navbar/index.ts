export { default, navLinks } from './Navbar'
