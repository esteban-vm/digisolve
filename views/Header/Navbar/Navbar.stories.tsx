import type { Meta, StoryObj } from '@storybook/react'
import { within, userEvent } from '@storybook/testing-library'
import { styled } from '@/utils/styles'
import Navbar from './Navbar'

const Wrapper = styled.div`
  width: 100%;
  height: 120px;
  background-color: var(--color-black);
`

export default {
  component: Navbar,
  title: 'Views/Header/Navbar',
  decorators: [
    (Story, context) => {
      return (
        <Wrapper>
          <Story {...context} />
        </Wrapper>
      )
    },
  ],
} as Meta<typeof Navbar>

export const NavbarExample: StoryObj<typeof Navbar> = {
  play({ canvasElement }) {
    const canvas = within(canvasElement)
    const links = canvas.getAllByRole('link')
    const logo = canvas.getByRole('img')
    for (const link of links) userEvent.click(link)
    userEvent.hover(logo)
    userEvent.unhover(logo)
  },
}
