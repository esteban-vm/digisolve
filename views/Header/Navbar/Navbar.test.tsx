import type { ReactTestRendererJSON } from 'react-test-renderer'
import { render, cleanup, create, screen } from '@/utils/tests'
import Navbar from './Navbar'

describe('ðŸ§ª <Navbar /> test cases:', () => {
  describe('there should be:', () => {
    beforeEach(() => {
      render(<Navbar />)
    })

    afterEach(cleanup)

    it('a logo', () => {
      const logo = screen.getByRole('img')
      expect(logo).toBeInTheDocument()
      expect(logo).toBeVisible()
      expect(logo).toHaveAccessibleName(/^digisolve logo$/i)
    })

    it('3 links', () => {
      const links = screen.getAllByRole('link')
      const texts = [/^home$/i, /^contact$/i, /^about$/i]
      expect(links).toHaveLength(3)

      for (let index = 0; index < links.length; index++) {
        const link = links[index]
        expect(link).toBeInTheDocument()
        expect(link).toBeVisible()
        expect(link).toHaveTextContent(texts[index])
      }
    })
  })

  describe('style tests:', () => {
    let navbarTree: ReactTestRendererJSON

    beforeEach(() => {
      navbarTree = create(<Navbar />).toJSON() as ReactTestRendererJSON
    })

    it('should render with correct styles', () => {
      expect(navbarTree).toMatchSnapshot()
    })

    it('should change color when hovered', () => {
      expect(navbarTree).toHaveStyleRule('border-bottom', '2px solid var(--color-primary)', { target: 'a:hover' })
    })
  })
})
