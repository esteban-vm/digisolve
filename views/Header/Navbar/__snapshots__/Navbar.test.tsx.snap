// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ðŸ§ª <Navbar /> test cases: style tests: should render with correct styles 1`] = `
.emotion-0 img {
  height: 120px;
  width: auto;
  float: left;
}

.emotion-0 ul {
  list-style: none;
  float: right;
  margin-top: 70px;
}

.emotion-0 ul li {
  display: inline-block;
  margin-right: 40px;
}

.emotion-0 ul li a:link,
.emotion-0 ul li a:visited {
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: uppercase;
  color: var(--color-white);
  outline: none;
  -webkit-transition: border-bottom 100ms;
  transition: border-bottom 100ms;
}

.emotion-0 ul li a:hover,
.emotion-0 ul li a:active {
  border-bottom: 2px solid var(--color-primary);
}

<nav
  className="emotion-0"
>
  <img
    alt="Digisolve logo"
    data-nimg="1"
    decoding="async"
    height={40}
    loading="lazy"
    onError={[Function]}
    onLoad={[Function]}
    src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
    srcSet="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
    style={
      {
        "color": "transparent",
      }
    }
    width={40}
  />
  <ul>
    <li>
      <a
        href="#"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        Home
      </a>
    </li>
    <li>
      <a
        href="#"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        Contact
      </a>
    </li>
    <li>
      <a
        href="#"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        About
      </a>
    </li>
  </ul>
</nav>
`;
